substrate/pkg/controller/substrate/infrastructure/natgateway.go

```diff
@@ -0,0 +1,127 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package infrastructure
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/awslabs/kubernetes-iteration-toolkit/substrate/pkg/apis/v1alpha1"
+	"github.com/awslabs/kubernetes-iteration-toolkit/substrate/pkg/utils/discovery"
+	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type NatGateway struct {
+	EC2 *ec2.EC2
+}
+
+func (n *NatGateway) Create(ctx context.Context, substrate *v1alpha1.Substrate) (reconcile.Result, error) {
+	if substrate.Status.Infrastructure.VPCID == nil ||
+		substrate.Status.Infrastructure.PrivateRouteTableID == nil ||
+		len(substrate.Status.Infrastructure.PublicSubnetIDs) == 0 ||
+		substrate.Status.Infrastructure.ElasticIpIDForNatGW == nil {
+		return reconcile.Result{Requeue: true}, nil
+	}
+	natGW, err := n.ensure(ctx, substrate)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if _, err := n.EC2.CreateRouteWithContext(ctx, &ec2.CreateRouteInput{
+		RouteTableId:         substrate.Status.Infrastructure.PrivateRouteTableID,
+		DestinationCidrBlock: aws.String("0.0.0.0/0"),
+		NatGatewayId:         natGW.NatGatewayId,
+	}); err != nil {
+		return reconcile.Result{}, fmt.Errorf("creating route for NAT gateway, %w", err)
+	} else {
+		logging.FromContext(ctx).Debugf("Ensured route for NAT gateway %s", aws.StringValue(natGW.NatGatewayId))
+	}
+	return reconcile.Result{}, nil
+}
+
+func (n *NatGateway) getNatGateway(ctx context.Context, substrate *v1alpha1.Substrate) (*ec2.NatGateway, error) {
+	output, err := n.EC2.DescribeNatGatewaysWithContext(ctx, &ec2.DescribeNatGatewaysInput{Filter: discovery.Filters(substrate, discovery.Name(substrate))})
+	if err != nil {
+		return nil, fmt.Errorf("describing nat-gateway, %w", err)
+	}
+	if len(output.NatGateways) == 0 {
+		return nil, nil
+	}
+	var result *ec2.NatGateway
+	for _, natgw := range output.NatGateways {
+		if aws.StringValue(natgw.State) == "deleting" || aws.StringValue(natgw.State) == "deleted" ||
+			aws.StringValue(natgw.State) == "failed" {
+			continue
+		}
+		if result != nil {
```

Comment:
Looks like we fail if we find more than one NATs in VPC ,  we can have a `Todo` for future (nice to have) to add atleast 2 NATs so we are immune to AZ failures for all clusters using KIT Environment :D

Reply:
yeah sure once we add HA to the control plane.